components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManagerV2.java

```diff
@@ -227,78 +195,53 @@ public RoleV2 getRole(String roleID, Map<String, Boolean> requiredAttributes)
-                for (UserBasicInfo userInfo : assignedUsers) {
-                    userInfo.getId();
-                    String userLocationURI = SCIMCommonUtils.getSCIMUserURL(userInfo.getId());
-                    User user = new User();
-                    user.setUserName(userInfo.getName());
-                    user.setId(userInfo.getId());
-                    user.setLocation(userLocationURI);
-                    scimRole.setUser(user);
-                }
-            }
+            setAssignedUsersForRole(scimRole, role.getUsers());
 
             // Set role's assigned userstore groups.
-            List<GroupBasicInfo> assignedUserstoreGroups = role.getGroups();
-            if (assignedUserstoreGroups != null) {
-                for (GroupBasicInfo groupInfo : assignedUserstoreGroups) {
-                    String groupId = groupInfo.getId();
-                    String groupLocationURI = SCIMCommonUtils.getSCIMGroupURL(groupId);
-                    Group group = new Group();
-                    group.setDisplayName(groupInfo.getName());
-                    group.setId(groupId);
-                    group.setLocation(groupLocationURI);
-                    scimRole.setGroup(group);
-                }
-            }
+            setRoleAssignedUserstoreGroups(scimRole, role.getGroups());
 
             // Set role's assigned idp groups.
-            List<IdpGroup> assignedIdpGroups = role.getIdpGroups();
-            if (assignedIdpGroups != null) {
-                for (IdpGroup idpGroup : assignedIdpGroups) {
-                    String idpGroupId = idpGroup.getGroupId();
-                    String idpGroupLocationURI = SCIMCommonUtils.getIdpGroupURL(idpGroup.getIdpId(), idpGroupId);
-                    Group group = new Group();
-                    group.setDisplayName(idpGroup.getGroupName());
-                    group.setId(idpGroupId);
-                    group.setLocation(idpGroupLocationURI);
-                    scimRole.setGroup(group);
-                }
-            }
+            getRoleAssignedIdpGroups(scimRole, role.getIdpGroups());
 
             // Set associated applications.
-            List<MultiValuedComplexType> associatedApps =
-                    convertAssociatedAppsToMultivaluedComplexType(role.getAssociatedApplications());
-            if (CollectionUtils.isNotEmpty(associatedApps)) {
-                scimRole.setAssociatedApplications(associatedApps);
+            setAssociatedPermissionsForRole(scimRole, role);
+
+            return scimRole;
+        } catch (IdentityRoleManagementException e) {
+            if (StringUtils.equals(ROLE_NOT_FOUND.getCode(), e.getErrorCode())) {
+                throw new NotFoundException(e.getMessage());
             }
+            throw new CharonException(String.format("Error occurred while getting the role: %s", roleID), e);
+        }
+    }
+
+    public RoleV2 getRoleV3(String roleID, Map<String, Boolean> requiredAttributes)
```

Comment:
getRoleV3 method in SCIMRoleManagerV2 class?

Reply:
Why do we need to have another get operation for v3 version? I guess v2 and v3 doesn't have any api changes.

Reply:
To ensure the correct version is assigned to the location value in the getResponse, we need to differentiate between v2 and v3 APIs. Using the existing method assigns the location version as v2, even when working with the v3 API, which leads to incorrect behavior. To address this issue and maintain accuracy, I have introduced a new method.

Related review comment: https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/645#discussion_r2150468638